### PR TITLE
Flow + GVP updates

### LINKED ISSUE #18: use loguru for logging instead of print statements

### LINKED ISSUE #19: fix relative imports
currently I use hacky sys path edits to make sure imports don't fail, fixed it for dataset.py; need to do it throughout the codebase

### LINKED ISSUE #23: Add a toggle for edge updates in the GVP

### LINKED ISSUE #51: Make `euler_integrate` returns the same type as `rk4_integrate`
`euler_integrate` returns `list[np.ndarray]` that needs further processing to match the `list[dict[str, np.ndarray]]` returned by `rk4_integrate`. This affects the `inference.py` script.

### PR DESCRIPTION
**Pull request by vratins**
- Add support for encoder-provided edge features (`pp_edge_attr`) in flow model, enabling GVP encoder to pass learned edge representations through the message passing layers
  - Add configurable `n_message_gvps` and `n_update_gvps` parameters to control GVP layer depth
  - Refactor `training_step` to separate gradient accumulation from optimizer step (caller now handles optimizer)
  - Use constants (`NUM_RBF, RBF_CUTOFF`) instead of hardcoded values throughout GVP module
  - Add comprehensive docstrings and standardize formatting

### CODE REVIEW

File: src/flow.py [hunk 1]
```diff
@@ -385,35 +520,41 @@ def compute_sigma_per_graph(data: HeteroData, device: torch.device) -> torch.Ten
...
         return sigma
 
     def training_step(
         self,
         batch: HeteroData,
-        optimizer: torch.optim.Optimizer,
-        grad_clip: float = 1.0,
         use_self_conditioning: bool = True,
+        accumulation_steps: int = 1,
     ) -> dict[str, float]:
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/flow.py]: Will be addressed in the train/inference PR (look at load_esm3 branch)

File: src/gvp.py [hunk 2]
```diff
@@ -9,12 +17,20 @@
 from torch_geometric.nn import HeteroConv, MessagePassing
 from torch_scatter import scatter_add
 
+from src.constants import NUM_RBF, RBF_CUTOFF
 from src.utils import rbf
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/gvp.py]: Addressed in another PR

**Review by marcuscollins — CHANGES_REQUESTED** [review]
I think these PRs need to be reorganized so that they don't break `main`. There are too many things here where it is hard to evaluate what's happening because they require changes in other places not included in this PR. 

I've put comments in here, so have a look, but try to isolate each functional change, like adding an argument to a method's signature, and make separate PRs for those. Make a separate PR for changes that are just formatting.

File: src/gvp_encoder.py [hunk 3]
```diff
@@ -20,16 +20,7 @@
...
-
-
-def _edge_vectors(pos: torch.Tensor, edge_index: torch.Tensor):
-    """Compute edge vectors and distances."""
-    src, dst = edge_index[0], edge_index[1]
-    vec = pos[dst] - pos[src]
-    rij = torch.linalg.norm(vec, dim=-1).clamp(min=1e-4)
-    r_hat = vec / rij[:, None]
-    return rij, r_hat
+from src.utils import compute_edge_geometry, rbf as _rbf
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/gvp_encoder.py]: I addressed this in #37, I could change it here as well, or take care of it in the merge.

File: src/utils.py [hunk 4]
```diff
@@ -55,6 +55,68 @@ def rbf(r: Tensor, num_gaussians: int = 16, cutoff: float = 8.0) -> Tensor:
...
+    Returns:
+        distances: (E,) edge distances
+        unit_vectors: (E, 3) unit displacement vectors from source to destination
+    """
+    src_idx, dst_idx = edge_index[0], edge_index[1]
+    pos_src = pos
+    if pos_dst is None:
+        pos_dst = pos
+    displacement = pos_dst[dst_idx] - pos_src[src_idx]
+    distances = torch.linalg.norm(displacement, dim=-1).clamp(min=clamp_min)
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/utils.py]: Yep, I will add more clarifications.

File: src/flow.py [hunk 5]
```diff
@@ -55,37 +76,54 @@ def __init__(
...
     ):
+        """
+        Initialize heterogeneous protein-water message passing module.
+
+        Args:
+            hidden_dims: (scalar_dim, vector_dim) hidden dimensions for GVP layers
+            rbf_dim: Number of radial basis functions for distance encoding
+            layers: Number of GVP message passing layers
+            drop_rate: Dropout rate for regularization
+            n_message_gvps: Number of GVPs in message function per edge type
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/flow.py]: Replying to both this comment and the one above ^ -- I believe the `n_<message/update>_gvps` is clearer than `layers` as they _are_ technically `GVP` blocks; renaming it to layers might lead to confusion with the external `layers` argument?

File: src/flow.py [hunk 6]
```diff
@@ -487,15 +618,8 @@ def training_step(
...
-        optimizer.zero_grad(set_to_none=True)
-        loss.backward()
-        if grad_clip > 0:
-            torch.nn.utils.clip_grad_norm_(
-                [p for p in self.model.parameters() if p.requires_grad],
-                max_norm=grad_clip,
-            )
-        optimizer.step()
+        # backward (scale loss for gradient accumulation)
+        (loss / accumulation_steps).backward()
```
Comment (marcuscollins) [inline src/flow.py]: why get rid of gradient clipping? And `optimizer.step()`? This feels like something isn't right, since I don't see that step taken anywhere else.

**Review by marcuscollins — CHANGES_REQUESTED** [review]
You get rid of some code in this PR (gradient clipping, stepping the optimizer) that worries me. I don't see where the optimizer.step() method is called, which can't be right, unless it was previously called in multiple places and you're removing one. Otherwise mostly requests for clarification.

File: src/gvp.py [hunk 7]
```diff
@@ -352,6 +410,19 @@ def __init__(
...
+        """
+        Initialize full GVP convolution layer with residual and feedforward.
+
+        Args:
+            node_dims: (n_scalar, n_vector) node feature dimensions
+            edge_dims: (n_scalar_edge, n_vector_edge) edge feature dimensions
+            n_message: Number of GVPs in message function
+            n_feedforward: Number of GVPs in feedforward function
+            drop_rate: Dropout probability
+            autoregressive: If True, use 'add' aggregation for masked decoding
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/gvp.py]: do mean `autoregressive` or the `n_message/n_feedforward`? Since this is the original gvp implementation (and works) I haven't changed much but I can add some clarifying comments

File: src/gvp.py [hunk 8]
```diff
@@ -638,9 +775,26 @@ def __init__(
...
+            use_dst_feats: If True, include destination features in messages
+            drop_rate: Dropout probability
+            aggr_edges: Per-edge aggregation method ('sum' or 'mean')
+            activations: (scalar_act, vector_act) activation functions
+            vector_gate: Whether to use vector gating
+        """
         super().__init__()
         self.s_dim, self.v_dim = s_dim, v_dim
         self.drop = Dropout(drop_rate)
+        GVP_ = functools.partial(GVP, activations=activations, vector_gate=vector_gate)
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/gvp.py]: I agree with the broader point of avoiding repeated code, and will add in an issue

**Review by marcuscollins — APPROVED** [review]
LGTM
